core-bundle/src/Filesystem/PermissionCheckingVirtualFilesystem.php

```diff
@@ -85,4 +185,25 @@ private function denyAccessUnlessGranted(string $attribute, Uuid|string $locatio
 
         throw $exception;
     }
+
+    private function canAccess(string $attribute, Uuid|string $location): bool
+    {
+        if (!$this->inner instanceof VirtualFilesystem) {
+            return false;
+        }
```

Comment:
This is because the `getPrefix()` method only exists in the `VirtualFilesystem` class and not in the `VirtualFilesystemInterface`. But why?

Reply:
Because prefixes and the mount manager are a concept of the Core implementation of the VFS.This cannot appear in the interface.

Reply:
Actually, we could remove the line. I already narrowed the constructor argument at a later point.

I wonder why phpstan did not pick this up. Maybe due to the trait. 🤔

Reply:
PHPStan (or PhpStorm) will pick up on it if you remove the line, because then it complains about the missing `getPrefix()` method on `$this->inner` (which is an instance of `VirtualFilesystemInterface`).

Reply:
Then maybe we should add a type annotation, wdyt?

Reply:
How is a type annotation better than `instanceof`?

Reply:
A type annotation is just a hint for phpstan to understand the code (which it could if it was smart, but isn't yet).

The current solution is s runtime operation, it costs a comparison which will always result in `false`.

Reply:
But then we should at least use a sophisticated annotation and not something like this:

```php
/** @var VirtualFilesystem $vfs */
$vfs = $this->inner;
```

Is there no way to "overwrite" the class member definition from the trait with an annotation?

Reply:
The easiest way would be to introduce a member `private readonly VirtualFilesystem $inner;` that will overwrite the trait's member.

Reply:
I don't think that works. Does it?

Reply:
Sure, it's the same like overwriting methods. :slightly_smiling_face: 
https://phpstan.org/r/37c086a3-ed00-44ad-8400-3048cfdda970

Reply:
Nvm, this does not work because of PHP enforcing covariance.

But this would: https://phpstan.org/r/77f81428-d2fe-4c98-9f87-f426131f6bdf - effectively making the trait a template, where T can be a narrowed type.

Reply:
See #7671